cub/agent/agent_batch_memcpy.cuh

```diff
@@ -0,0 +1,1126 @@
+  }
+
+  /**
+   * @brief Computes the histogram over the number of buffers belonging to each of the three
+   * size-classes (TLEV, WLEV, BLEV).
+   */
+  template <typename VectorizedSizeClassCounterT>
+  __device__ __forceinline__ void
+  GetBufferSizeClassHistogram(const BufferSizeT (&buffer_sizes)[BUFFERS_PER_THREAD],
+                              VectorizedSizeClassCounterT &vectorized_counters)
+  {
+#pragma unroll
+    for (uint32_t i = 0; i < BUFFERS_PER_THREAD; i++)
+    {
+      // Whether to increment ANY of the buffer size classes at all
+      const uint32_t increment = buffer_sizes[i] > 0 ? 1U : 0U;
+      // Identify the buffer's size class
+      uint32_t buffer_size_class = 0;
+      buffer_size_class += buffer_sizes[i] > WARP_LEVEL_THRESHOLD ? 1U : 0U;
+      buffer_size_class += buffer_sizes[i] > BLOCK_LEVEL_THRESHOLD ? 1U : 0U;
+
+      // Increment the count of the respective size class
+      vectorized_counters.Add(buffer_size_class, increment);
+    }
+  }
+
+  /**
+   * @brief Scatters the buffers into the respective buffer's size-class partition.
+   */
+  template <typename VectorizedSizeClassCounterT>
+  __device__ __forceinline__ void
+  PartitionBuffersBySize(const BufferSizeT (&buffer_sizes)[BUFFERS_PER_THREAD],
+                         VectorizedSizeClassCounterT &vectorized_offsets,
+                         BufferTuple (&buffers_by_size_class)[BUFFERS_PER_BLOCK])
+  {
+    // If we intend to perform a stable partitioning, the thread's buffer are in a blocked
+    // arrangement, otherwise they are in a striped arrangement
+    BlockBufferOffsetT buffer_id = BUFFER_STABLE_PARTITION ? (BUFFERS_PER_THREAD * threadIdx.x)
+                                                           : (threadIdx.x);
+    constexpr BlockBufferOffsetT BUFFER_STRIDE = BUFFER_STABLE_PARTITION
+                                                   ? static_cast<BlockBufferOffsetT>(1)
+                                                   : static_cast<BlockBufferOffsetT>(BLOCK_THREADS);
+
+#pragma unroll
+    for (uint32_t i = 0; i < BUFFERS_PER_THREAD; i++)
+    {
+      if (buffer_sizes[i] > 0)
+      {
+        uint32_t buffer_size_class = 0;
+        buffer_size_class += buffer_sizes[i] > WARP_LEVEL_THRESHOLD ? 1U : 0U;
+        buffer_size_class += buffer_sizes[i] > BLOCK_LEVEL_THRESHOLD ? 1U : 0U;
+        const uint32_t write_offset         = vectorized_offsets.Get(buffer_size_class);
+        buffers_by_size_class[write_offset] = {static_cast<TLevBufferSizeT>(buffer_sizes[i]),
+                                               buffer_id};
+        vectorized_offsets.Add(buffer_size_class, 1U);
+      }
+      buffer_id += BUFFER_STRIDE;
+    }
+  }
+
+  /**
+   * @brief Read in all the buffers that require block-level collaboration and put them to a queue
+   * that will get picked up in a separate, subsequent kernel.
+   */
+  template <typename TileBufferOffsetT, typename TileOffsetT>
```

Comment:
Isn't this method used with a single type of both `TileBufferOffsetT` and `TileOffsetT`?

Reply:
Agreed. Removed the template parameters and made the respective parameters `BufferOffsetT`